codalab/worker/local_run/local_dependency_manager.py

```diff
@@ -87,45 +89,79 @@ def _save_state(self):
+
         dependencies = {}
         dependency_locks = {}
-        paths = set()
-        for dep_key, dep_state in state['dependencies'].items():
-            full_path = os.path.join(self.dependencies_dir, dep_state.path)
-            if os.path.exists(full_path):
-                dependencies[dep_key] = dep_state
-                dependency_locks[dep_key] = threading.RLock()
-            else:
-                logger.info(
-                    "Dependency {} in loaded state but its path {} doesn't exist in the filesystem".format(
-                        dep_key, full_path
-                    )
-                )
-            if dep_state.path not in state['paths']:
-                state['paths'].add(dep_state.path)
-                logger.info(
-                    "Dependency {} in loaded state but its path {} is not in the loaded paths {}".format(
-                        dep_key, dep_state.path, state['paths']
-                    )
-                )
-        for path in state['paths']:
-            full_path = os.path.join(self.dependencies_dir, path)
-            if os.path.exists(full_path):
-                paths.add(path)
-            else:
-                logger.info(
-                    "Path {} in loaded state but doesn't exist in the filesystem".format(full_path)
-                )
+
+        for dep, dep_state in state['dependencies'].items():
+            dependencies[dep] = dep_state
+            dependency_locks[dep] = threading.RLock()
 
         with self._global_lock, self._paths_lock:
             self._dependencies = dependencies
             self._dependency_locks = dependency_locks
-            self._paths = paths
+            self._paths = state['paths']
+
         logger.info(
-            '{} dependencies, {} paths in cache.'.format(len(self._dependencies), len(self._paths))
+            'Loaded {} dependencies, {} paths from cache.'.format(
+                len(self._dependencies), len(self._paths)
+            )
+        )
+
+    def _sync_state(self):
+        """
+        Remove any path that exists in the local file system but not in dependency-state.json file and
+        save the synchronized dependency state and loaded paths back to dependency-state.json file
+        :return:
+        """
+        # Get all the dependency directories in local file system under self.dependencies_dir
+        local_directories = set(os.listdir(self.dependencies_dir))
+
+        # Remove the orphaned dependencies from self._dependencies and
+        # self._dependency_locks if they don't exist in local file system
+        dependencies_to_remove = [
+            dep
+            for dep, dep_state in self._dependencies.items()
+            if dep_state.path not in local_directories
+        ]
+        for dep in dependencies_to_remove:
+            logger.info(
+                "Dependency {} in loaded state but its path {} doesn't exist in the file system. "
```

Comment:
Replace 'loaded state' with 'dependency state' (this comment applies more broadly)